erpnext/accounts/doctype/payment_entry/payment_entry.py

```diff
@@ -444,6 +444,7 @@ def set_missing_values(self):
 				self.party_name = frappe.db.get_value(self.party_type, self.party, "name")
 
 		if self.party:
+			set_contact_details(self, party=frappe._dict({"name": self.party}), party_type=self.party_type)
```

Comment:
```suggestion
			if not self.contact_person:
				set_contact_details(self, party=frappe._dict({"name": self.party}), party_type=self.party_type)
```

Otherwise this will always overwrite a manually selected contact with the primary contact.